File: bindings/julia/src/Ssh.jl
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
#
# Julia bindings for the proven-ssh-bastion protocol (SSH bastion host).
#
# Wraps the C-ABI functions from protocols/proven-ssh-bastion/ffi/zig/src/ssh_bastion.zig
# via ccall into libproven_ssh_bastion.so.

module Ssh

using ..ProvenServers: check_status, check_slot, SlotId

export BastionState, KexMethod, AuthMethod, ChannelType, ChannelState,
       DisconnectReason,
       abi_version, create, destroy, get_state, complete_kex, authenticate,
       open_channel, confirm_channel, close_channel, channel_count,
       rekey, disconnect, can_transition, is_recording

const LIB = "libproven_ssh_bastion"

# --------------------------------------------------------------------------
# Enumeration types matching Idris2 ABI
# --------------------------------------------------------------------------

"""SSH bastion session states."""
@enum BastionState::UInt8 begin
    STATE_INIT          = 0
    STATE_KEX           = 1
    STATE_AUTH          = 2
    STATE_AUTHENTICATED = 3
    STATE_CHANNEL_OPEN  = 4
    STATE_REKEY         = 5
    STATE_DISCONNECTED  = 6
end

"""SSH key exchange methods."""
@enum KexMethod::UInt8 begin
    KEX_CURVE25519_SHA256   = 0
    KEX_ECDH_SHA2_NISTP256  = 1
    KEX_ECDH_SHA2_NISTP384  = 2
    KEX_DH_GROUP14_SHA256   = 3
    KEX_DH_GROUP16_SHA512   = 4
end

"""SSH authentication methods."""
@enum AuthMethod::UInt8 begin
    AUTH_PUBLIC_KEY           = 0
    AUTH_PASSWORD             = 1
    AUTH_KEYBOARD_INTERACTIVE = 2
    AUTH_CERTIFICATE          = 3
end

"""SSH channel types."""
@enum ChannelType::UInt8 begin
    CHANNEL_SESSION         = 0
    CHANNEL_DIRECT_TCPIP    = 1
    CHANNEL_FORWARDED_TCPIP = 2
    CHANNEL_X11             = 3
end

"""SSH channel states."""
@enum ChannelState::UInt8 begin
    CH_OPENING             = 0
    CH_OPEN                = 1
    CH_HALF_CLOSED_LOCAL   = 2
    CH_HALF_CLOSED_REMOTE  = 3
    CH_CLOSED              = 4
end

"""SSH disconnect reasons."""
@enum DisconnectReason::UInt8 begin
    REASON_BY_APPLICATION         = 0
    REASON_TOO_MANY_AUTH_FAILURES = 1
    REASON_PROTOCOL_ERROR         = 2
    REASON_TIMEOUT                = 3
end

# --------------------------------------------------------------------------
# ccall declarations
# --------------------------------------------------------------------------

"""Return the ABI version of the linked libproven_ssh_bastion."""
function abi_version()::UInt32
    ccall((:ssh_bastion_abi_version, LIB), UInt32, ())
end

"""
    create(kex::KexMethod, auth::AuthMethod) -> SlotId

Create a new SSH bastion session. Throws on pool exhaustion.
"""
function create(kex::KexMethod, auth::AuthMethod)::SlotId
    check_slot(ccall((:ssh_bastion_create, LIB), Cint,
                     (UInt8, UInt8), UInt8(kex), UInt8(auth)))
end

"""
    destroy(slot::SlotId)

Release the given SSH bastion context slot.
"""
function destroy(slot::SlotId)::Nothing
    ccall((:ssh_bastion_destroy, LIB), Cvoid, (Cint,), slot)
    nothing
end

"""
    get_state(slot::SlotId) -> BastionState

Get the current SSH bastion session state.
"""
function get_state(slot::SlotId)::BastionState
    BastionState(ccall((:ssh_bastion_state, LIB), UInt8, (Cint,), slot))
end

"""
    complete_kex(slot::SlotId)

Complete key exchange. Throws on invalid state.
"""
function complete_kex(slot::SlotId)::Nothing
    check_status(ccall((:ssh_bastion_complete_kex, LIB), UInt8, (Cint,), slot))
end

"""
    authenticate(slot::SlotId, user_len::UInt16)

Authenticate user. Throws on invalid state.
"""
function authenticate(slot::SlotId, user_len::UInt16)::Nothing
    check_status(ccall((:ssh_bastion_authenticate, LIB), UInt8,
                       (Cint, UInt16), slot, user_len))
end

"""
    open_channel(slot::SlotId, ch_type::ChannelType) -> Cint

Open a new channel. Returns channel ID. Throws on pool exhaustion.
"""
function open_channel(slot::SlotId, ch_type::ChannelType)::Cint
    check_slot(ccall((:ssh_bastion_open_channel, LIB), Cint,
                     (Cint, UInt8), slot, UInt8(ch_type)))
end

"""
    confirm_channel(slot::SlotId, ch_id::UInt8)

Confirm an open channel. Throws on invalid state.
"""
function confirm_channel(slot::SlotId, ch_id::UInt8)::Nothing
    check_status(ccall((:ssh_bastion_confirm_channel, LIB), UInt8,
                       (Cint, UInt8), slot, ch_id))
end

"""
    close_channel(slot::SlotId, ch_id::UInt8)

Close a channel. Throws on invalid state.
"""
function close_channel(slot::SlotId, ch_id::UInt8)::Nothing
    check_status(ccall((:ssh_bastion_close_channel, LIB), UInt8,
                       (Cint, UInt8), slot, ch_id))
end

"""
    channel_count(slot::SlotId) -> UInt8

Get the number of open channels.
"""
function channel_count(slot::SlotId)::UInt8
    ccall((:ssh_bastion_channel_count, LIB), UInt8, (Cint,), slot)
end

"""
    rekey(slot::SlotId)

Initiate re-keying. Throws on invalid state.
"""
function rekey(slot::SlotId)::Nothing
    check_status(ccall((:ssh_bastion_rekey, LIB), UInt8, (Cint,), slot))
end

"""
    disconnect(slot::SlotId, reason::DisconnectReason)

Disconnect with the given reason. Throws on invalid state.
"""
function disconnect(slot::SlotId, reason::DisconnectReason)::Nothing
    check_status(ccall((:ssh_bastion_disconnect, LIB), UInt8,
                       (Cint, UInt8), slot, UInt8(reason)))
end

"""
    can_transition(from::BastionState, to::BastionState) -> Bool

Check whether an SSH bastion state transition is valid.
"""
function can_transition(from::BastionState, to::BastionState)::Bool
    ccall((:ssh_bastion_can_transition, LIB), UInt8,
          (UInt8, UInt8), UInt8(from), UInt8(to)) == 0x01
end

"""
    is_recording(slot::SlotId) -> Bool

Check if session recording is enabled.
"""
function is_recording(slot::SlotId)::Bool
    ccall((:ssh_bastion_is_recording, LIB), UInt8, (Cint,), slot) == 0x01
end

end # module Ssh

File: bindings/dart/lib/src/ssh.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
/// SSH Bastion protocol bindings for proven-servers.
///
/// Mirrors the Idris2 module `SshBastionABI.Types`. All discriminant
/// values match the Idris2 ABI tag definitions exactly.
///
/// See `protocols/proven-ssh-bastion/src/` for the Idris2 definitions.

import 'dart:ffi';

import 'error.dart';
import 'ffi.dart';

// ---------------------------------------------------------------------------
// SSH Constants
// ---------------------------------------------------------------------------

/// Standard SSH port (RFC 4253).
const int sshPort = 22;

// ---------------------------------------------------------------------------
// SshMessageType (tags 0-7)
// ---------------------------------------------------------------------------

/// SSH message types.
///
/// Matches `SshMessageType` in `SshBastionABI.Types`.
enum SshMessageType {
  kexinit(0),
  newkeys(1),
  serviceRequest(2),
  userauthRequest(3),
  channelOpen(4),
  channelData(5),
  channelClose(6),
  disconnect(7);

  final int tag;
  const SshMessageType(this.tag);

  static SshMessageType? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// AuthMethod (tags 0-3)
// ---------------------------------------------------------------------------

/// SSH authentication methods.
enum SshAuthMethod {
  password(0),
  publicKey(1),
  keyboardInteractive(2),
  none(3);

  final int tag;
  const SshAuthMethod(this.tag);

  static SshAuthMethod? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// KexMethod (tags 0-5)
// ---------------------------------------------------------------------------

/// SSH key exchange methods.
enum SshKexMethod {
  curve25519Sha256(0),
  ecdhSha2Nistp256(1),
  ecdhSha2Nistp384(2),
  ecdhSha2Nistp521(3),
  diffieHellmanGroup14Sha256(4),
  diffieHellmanGroup16Sha512(5);

  final int tag;
  const SshKexMethod(this.tag);

  static SshKexMethod? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// ChannelType (tags 0-3)
// ---------------------------------------------------------------------------

/// SSH channel types.
enum SshChannelType {
  session(0),
  directTcpip(1),
  forwardedTcpip(2),
  x11(3);

  final int tag;
  const SshChannelType(this.tag);

  static SshChannelType? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// BastionState (tags 0-5)
// ---------------------------------------------------------------------------

/// SSH bastion connection state machine.
enum SshBastionState {
  initial(0),
  kexInProgress(1),
  authenticated(2),
  channelOpen(3),
  dataTransfer(4),
  disconnected(5);

  final int tag;
  const SshBastionState(this.tag);

  static SshBastionState? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// HostKeyAlgorithm (tags 0-3)
// ---------------------------------------------------------------------------

/// SSH host key algorithms.
enum SshHostKeyAlgorithm {
  sshEd25519(0),
  ecdsaSha2Nistp256(1),
  rsaSha2_256(2),
  rsaSha2_512(3);

  final int tag;
  const SshHostKeyAlgorithm(this.tag);

  static SshHostKeyAlgorithm? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// CipherAlgorithm (tags 0-5)
// ---------------------------------------------------------------------------

/// SSH symmetric cipher algorithms.
enum SshCipherAlgorithm {
  chacha20Poly1305(0),
  aes256Gcm(1),
  aes128Gcm(2),
  aes256Ctr(3),
  aes192Ctr(4),
  aes128Ctr(5);

  final int tag;
  const SshCipherAlgorithm(this.tag);

  static SshCipherAlgorithm? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// DisconnectReason (tags 0-11)
// ---------------------------------------------------------------------------

/// SSH disconnect reason codes (RFC 4253 Section 11.1).
enum SshDisconnectReason {
  hostNotAllowed(0),
  protocolError(1),
  keyExchangeFailed(2),
  reserved(3),
  macError(4),
  compressionError(5),
  serviceNotAvailable(6),
  protocolVersion(7),
  hostKeyNotVerifiable(8),
  connectionLost(9),
  byApplication(10),
  tooManyConnections(11);

  final int tag;
  const SshDisconnectReason(this.tag);

  static SshDisconnectReason? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// ChannelOpenFailure (tags 0-3)
// ---------------------------------------------------------------------------

/// SSH channel open failure reasons.
enum SshChannelOpenFailure {
  administrativelyProhibited(0),
  connectFailed(1),
  unknownChannelType(2),
  resourceShortage(3);

  final int tag;
  const SshChannelOpenFailure(this.tag);

  static SshChannelOpenFailure? fromTag(int tag) {
    if (tag >= 0 && tag < values.length) return values[tag];
    return null;
  }
}

// ---------------------------------------------------------------------------
// SshContext — safe wrapper with dispose pattern
// ---------------------------------------------------------------------------

/// An SSH context slot in the Zig FFI pool.
///
/// Wraps the `ssh_*` C functions with automatic resource cleanup.
class SshContext {
  final ProvenFfi _ffi;
  final int _slot;
  bool _disposed = false;

  late final _destroy = _ffi.lookupDestroyContext('ssh_destroy_context');
  late final _getState = _ffi.lookupGetTag('ssh_get_state');
  late final _setAuthMethod = _ffi.lookupSetTag('ssh_set_auth_method');
  late final _setKexMethod = _ffi.lookupSetTag('ssh_set_kex_method');
  late final _getMessageType = _ffi.lookupGetTag('ssh_get_message_type');

  SshContext._(this._ffi, this._slot);

  /// Create a new SSH context.
  ///
  /// Throws [ProvenError] if the pool is exhausted.
  factory SshContext.create(ProvenFfi ffi) {
    final create = ffi.lookupCreateContext('ssh_create_context');
    final slot = ProvenError.checkSlot(create());
    return SshContext._(ffi, slot);
  }

  /// Release the context slot back to the pool.
  void dispose() {
    if (!_disposed) {
      _destroy(_slot);
      _disposed = true;
    }
  }

  void _checkDisposed() {
    if (_disposed) throw const ProvenError('context already disposed');
  }

  /// Get the current bastion state.
  SshBastionState? getState() {
    _checkDisposed();
    return SshBastionState.fromTag(_getState(_slot));
  }

  /// Set the authentication method.
  ///
  /// Throws [ProvenError] on invalid parameter.
  void setAuthMethod(SshAuthMethod method) {
    _checkDisposed();
    ProvenError.checkParamStatus(_setAuthMethod(_slot, method.tag));
  }

  /// Set the key exchange method.
  ///
  /// Throws [ProvenError] on invalid parameter.
  void setKexMethod(SshKexMethod method) {
    _checkDisposed();
    ProvenError.checkParamStatus(_setKexMethod(_slot, method.tag));
  }

  /// Get the last processed message type.
  SshMessageType? getMessageType() {
    _checkDisposed();
    return SshMessageType.fromTag(_getMessageType(_slot));
  }
}
